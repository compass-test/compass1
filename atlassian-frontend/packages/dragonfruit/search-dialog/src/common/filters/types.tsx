import { Products } from '../product-context';

export enum FilterOptionSource {
  /**
   * This means the result came from collab graph
   */
  COLLABORATION_GRAPH = 'COLLABORATION_GRAPH',
  /**
   * This means the result came from smart user picker
   */
  SMART_USER_PICKER = 'SMART_USER_PICKER',
  /**
   * This means the result came from a recents (bootstrap) query
   */
  RECENT = 'RECENT',
  /**
   * This means the result came from an explicit search
   */
  SEARCH = 'SEARCH',
  /**
   * This means the result is derived from the current user
   */
  CURRENT_USER = 'CURRENT_USER',
  /**
   * This means the result came from an external source passed into the component
   */
  EXTERNAL = 'EXTERNAL',
  /**
   * This means the result is static without any user-generation
   */
  STATIC = 'STATIC',
}

export interface FilterOption {
  id: string;
  isChecked: boolean;
  isVisible: boolean;
  filterSource: FilterOptionSource;
}

export interface SiteFilterOption extends FilterOption {
  product: Products;
  avatarUrl: string;
  cloudId: string;
  siteName: string;
  siteUrl: string;
}
