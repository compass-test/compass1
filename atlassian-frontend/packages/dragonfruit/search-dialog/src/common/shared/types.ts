export interface ResultState<T> {
  isLoading: boolean;
  results: T | null;
}

export enum ActionSubjectId {
  preQuerySearchResults = 'preQuerySearchResults',
  postQuerySearchResults = 'postQuerySearchResults',
  cachedResults = 'cachedResults',
}
