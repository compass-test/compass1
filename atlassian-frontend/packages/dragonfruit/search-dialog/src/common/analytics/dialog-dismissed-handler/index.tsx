export { DialogDismissedHandler } from './dialog-dismissed-handler';
