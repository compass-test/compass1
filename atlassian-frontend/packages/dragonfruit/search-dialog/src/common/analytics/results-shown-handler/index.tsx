export { default as SearchResultsShownHandler } from './results-shown-handler';
