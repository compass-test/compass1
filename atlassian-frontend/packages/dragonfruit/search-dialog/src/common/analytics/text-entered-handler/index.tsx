export { default as TextEnteredHandler } from './text-entered-handler';
