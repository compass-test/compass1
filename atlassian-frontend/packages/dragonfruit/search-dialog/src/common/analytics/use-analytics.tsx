import React from 'react';
import { useAnalyticsEvents } from '@atlaskit/analytics-next';
import { DEFAULT_GAS_CHANNEL, LimitedGasPayload } from './events';

export const useAnalytics = () => {
  const { createAnalyticsEvent } = useAnalyticsEvents();

  return {
    fireAnalyticsEvent: React.useCallback(
      (payload: LimitedGasPayload) =>
        createAnalyticsEvent(payload).fire(DEFAULT_GAS_CHANNEL),
      [createAnalyticsEvent],
    ),
  };
};
