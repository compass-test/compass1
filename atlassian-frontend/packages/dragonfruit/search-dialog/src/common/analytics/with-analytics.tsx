import React from 'react';
import {
  withAnalyticsEvents,
  WithAnalyticsEventsProps,
} from '@atlaskit/analytics-next';
import {
  DEFAULT_GAS_CHANNEL,
  LimitedGasPayload,
  LimitedGasScreenEventPayload,
} from './events';

type Callback = (
  ...args: any[]
) => LimitedGasPayload | LimitedGasScreenEventPayload;

export function withAnalytics<T extends Record<string, Callback>>(
  analytics: T,
) {
  type AnalyticsType = typeof analytics;

  return <U extends {}>(Component: React.ComponentType<U & AnalyticsType>) => {
    type WrapperComponentProps = Omit<U, keyof AnalyticsType> &
      WithAnalyticsEventsProps;

    class WrappedComponent extends React.Component<WrapperComponentProps> {
      wrappedAnalyticsCallback = Object.keys(analytics).reduce(
        (acc, key) => ({
          ...acc,
          // We forward all args to the corresponding callback regardless of type
          [key]: (...args: any[]) => {
            this.props.createAnalyticsEvent &&
              this.props
                .createAnalyticsEvent(analytics[key](...args))
                .fire(DEFAULT_GAS_CHANNEL);
          },
        }),
        {} as Record<string, (...args: any[]) => void>,
      );

      render() {
        const { createAnalyticsEvent, ...rest } = this.props;

        // There is NOT typesafety for the next line as typescript cannot infer types correctly with spread operators.
        // This is just going to have to go without type safety for the next line.
        return (
          <Component {...rest} {...(this.wrappedAnalyticsCallback as any)} />
        );
      }
    }

    return withAnalyticsEvents()(WrappedComponent);
  };
}
