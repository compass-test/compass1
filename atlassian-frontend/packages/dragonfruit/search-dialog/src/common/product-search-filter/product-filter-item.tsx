import React, { FunctionComponent, useCallback } from 'react';
import { FilterItem, FilterItemProps } from '@atlassian/search-dialog';
import { useAnalytics, onFilterSelect, onFilterUnselect } from '../analytics';
import { FilterOptionSource } from '../../common/filters/types';
import { ProductFilterType } from '../../common/clients/common-types';

export type OnChangeHandler = (
  value: string,
  type: ProductFilterType,
  isChecked: boolean,
) => any;

interface Props extends Omit<FilterItemProps, 'onChange'> {
  filterType: ProductFilterType;
  LabelComponent: React.ElementType;
  onChange?: OnChangeHandler;
  defaultChecked?: boolean;
  filterSource: FilterOptionSource;
  isDisabled?: boolean;
}

export const ProductFilterItem: FunctionComponent<Props> = ({
  onChange,
  filterType,
  filterSource,
  isDisabled = false,
  ...rest
}: Props) => {
  const { fireAnalyticsEvent } = useAnalytics();
  const onChangeCallback = useCallback(
    (id: string, checked: boolean) => {
      if (checked) {
        fireAnalyticsEvent(
          onFilterSelect({
            filterId: id,
            filterSource,
            filterType,
          }),
        );
      } else {
        fireAnalyticsEvent(
          onFilterUnselect({
            filterId: id,
            filterSource,
            filterType,
          }),
        );
      }

      onChange?.(id, filterType, checked);
    },
    [onChange, fireAnalyticsEvent, filterType, filterSource],
  );

  return (
    <FilterItem {...rest} onChange={onChangeCallback} isDisabled={isDisabled} />
  );
};
