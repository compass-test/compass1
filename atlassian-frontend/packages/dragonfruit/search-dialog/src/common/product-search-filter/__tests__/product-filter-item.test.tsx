import React from 'react';
import { mount } from 'enzyme';
import { ProductFilterItem } from '../product-filter-item';
import { onFilterSelect, onFilterUnselect } from '../../analytics';
import { FilterColLabelText, FilterItem } from '@atlassian/search-dialog';
import { FilterType } from '../../../confluence/clients';
import { FilterOptionSource } from '../../../common/filters/types';

jest.mock('@atlassian/search-dialog', () => ({ FilterItem: () => <div /> }));
jest.mock('../../analytics', () =>
  Object.assign({}, jest.requireActual('../../analytics'), {
    useAnalytics: () => ({
      fireAnalyticsEvent: () => {},
    }),
    onFilterSelect: jest.fn(),
    onFilterUnselect: jest.fn(),
  }),
);

describe('<ProductFilterItem />', () => {
  const Icon = () => <div />;

  const commonProps = {
    filterType: 'space' as FilterType,
    icon: <Icon />,
    value: 'some value',
    label: 'some label',
    filterSource: FilterOptionSource.COLLABORATION_GRAPH,
    LabelComponent: FilterColLabelText,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    const wrapper = mount(<ProductFilterItem {...commonProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('fires analytic event when filter is checked', () => {
    const wrapper = mount(<ProductFilterItem {...commonProps} />);
    wrapper.find(FilterItem).prop('onChange')!('some value', true);
    expect(onFilterSelect).toBeCalledWith({
      filterId: 'some value',
      filterType: commonProps.filterType,
      filterSource: FilterOptionSource.COLLABORATION_GRAPH,
    });
  });

  it('fires analytic event when filter is unchecked', () => {
    const wrapper = mount(<ProductFilterItem {...commonProps} />);
    wrapper.find(FilterItem).prop('onChange')!('some value', false);
    expect(onFilterUnselect).toBeCalledWith({
      filterId: 'some value',
      filterType: commonProps.filterType,
      filterSource: FilterOptionSource.COLLABORATION_GRAPH,
    });
  });
});
