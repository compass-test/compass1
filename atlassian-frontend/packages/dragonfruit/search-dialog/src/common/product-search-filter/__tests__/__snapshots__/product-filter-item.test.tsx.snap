// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductFilterItem /> matches snapshot 1`] = `
<ProductFilterItem
  filterSource="COLLABORATION_GRAPH"
  filterType="space"
  icon={<Icon />}
  label="some label"
  value="some value"
>
  <FilterItem
    icon={<Icon />}
    isDisabled={false}
    label="some label"
    onChange={[Function]}
    value="some value"
  >
    <div />
  </FilterItem>
</ProductFilterItem>
`;
